File: ruby/shipping_address.rb
require "rest_client"
require "json"

module PayFabric
  # 
  #  module provides methods for work with 
  #  `Sandbox` submodule provides methods for work with sandbox api
  #  `Live` submodule provides methods for work with production api
  module ShippingAdress
    module Share #:nodoc:
      extend self
      
      def retrieve_by_customer(url, device_id, password) #:nodoc:
        response = RestClient.get(url, 
          PayFabric::headers(device_id, password, url.include?("sandbox")))
        JSON.parse(response.body)
      end

      def retrieve(url, device_id, password) #:nodoc:
        response = RestClient.get(url, 
          PayFabric::headers(device_id, password, url.include?("sandbox")))
        JSON.parse(response.body)
      end
    end

    module Sandbox
      extend self 

      #  ## Retrieve Shipping Addresses By Customer
      #  [Info](https://github.com/PayFabric/APIs/wiki#retrieve-shipping-addresses-by-customer)
      #  
      #  +customer+ - unique name of custumer
      #  
      #  Example
      #    
      #    customer = "abc" 
      #    array = PayFabric::ShippingAdress::Sandbox::retrieve_by_customer(device_id, password, customer)
      #    array # => [{"ID" => "123"}, {"ID" => "124"}, {"ID" => "125"}]
      #    
      #  @return [Array with Adresses](https://github.com/PayFabric/APIs/wiki/API-Objects#address)
      #
      def retrieve_by_customer(device_id, password, customer)
        url = "https://sandbox.payfabric.com/rest/v1/api/addresses/#{customer}"
        Share::retrieve_by_customer(url, device_id, password)
      end

      #  ## Retrieve Shipping Address By Id
      #  [Info](https://github.com/PayFabric/APIs/wiki#retrieve-shipping-address-by-id)
      #  
      #  +id+ - customer id
      #  
      #  Example
      #   
      #    id = "10" 
      #    hash = PayFabric::ShippingAdress::Sandbox::retrieve(device_id, password, id)
      #    hash # => {"ID" => "123"}
      #
      #  @return [Adress Hash](https://github.com/PayFabric/APIs/wiki/API-Objects#address)
      #
      def retrieve(device_id, password, id)
        url = "https://sandbox.payfabric.com/rest/v1/api/address/#{id}"
        Share::retrieve(url, device_id, password)
      end
    end

    module Live
      extend self

      def retrieve_by_customer(device_id, password, customer)
        url = "https://payfabric.com/rest/v1/api/addresses/#{customer}"
        Share::retrieve_by_customer(url, device_id, password)
      end

      def retrieve(device_id, password, id)
        url = "https://payfabric.com/rest/v1/api/address/#{id}"
        Share::retrieve(url, device_id, password)
      end
    end
  end

end

File: python/ShippingAddress.py
# =====================================================================
# THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
#  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
#  PARTICULAR PURPOSE.
# =====================================================================

import requests
from Token import Token


class ShippingAddress:
    """ ShippingAddress methods
    """

    def Retrieve(self, addressId):
        """ Retrieve a shipping address record by id

            Keywords:
                addressId - Address GUID
        """

        r = requests.get(url='https://sandbox.payfabric.com/rest/v1/api/address/' + addressId,
                         headers={
                             'Content-Type': 'application/json; charset=utf-8',
                             'authorization': Token().Create()
                         })

        print r.status_code, r.text

        # Sample response
        # ------------------------------------------------------
        # Response text is an address object with json format
        # Go to https://github.com/PayFabric/APIs/wiki/API-Objects#address for more details about address object.
        # ------------------------------------------------------
        #
        return r.json()

    def RetrieveByCustomer(self, customer):
        """ Retrieve all shipping addresses by customer

        :param customer: Customer unique name
        :return: all shipping addresses
        """

        r = requests.get(url='https://sandbox.payfabric.com/rest/v1/api/addresses/' + customer,
                         headers={
                             'Content-Type': 'application/json; charset=utf-8',
                             'authorization': Token().Create()
                         })

        print r.status_code, r.text

        # Sample response
        # ------------------------------------------------------
        # Response text is an array of address object with json format
        # Go to https://github.com/PayFabric/APIs/wiki/API-Objects#address for more details about address object.
        # ------------------------------------------------------
        #
        return r.json();